### PR TITLE
bsunanda:Run2-hcx83 Update Geometry to handle minimum depth > 1 and max phi > 72

### PR DESCRIPTION
**Pull request by bsunanda**

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by kpedro88** [comment]
@bsunanda still needs a tweak... it knows minDepth=2 for ieta=17, but now it thinks maxDepth=4.

**Comment by bsunanda** [comment]
@kpedro88 Can you give me the matrix # which can reproduce the problem

**Comment by kpedro88** [comment]
@bsunanda It's a 2017dev workflow, so there's not really a good runTheMatrix command right now. Here's a script I've been using with the latest IBs:
```
#!/bin/bash

cmsDriver.py SinglePiE50HCAL_pythia8_cfi  --conditions auto:phase1_2017_design -n 10 --era Run2_2017 --geometry "Configuration.Geometry.GeometryExtended2017dev_cff,Configuration.Geometry.GeometryExtended2017devReco_cff" --eventcontent FEVTDEBUG -s GEN,SIM --datatier GEN-SIM --beamspot Realistic50ns13TeVCollision --fileout file:step1.root --customise SLHCUpgradeSimulations/Configuration/HCalCustoms.customise_Hcal2017 > step1_SinglePiE50_UP15+SinglePiE50_UP15+DIGIUP15+RECOUP15+HARVESTUP15.log  2>&1

cmsDriver.py step2  --conditions auto:phase1_2017_design -s DIGI:pdigi_valid,L1,DIGI2RAW,HLT:@relval25ns --datatier GEN-SIM-DIGI-RAW-HLTDEBUG -n 10 --era Run2_2017 --geometry "Configuration.Geometry.GeometryExtended2017dev_cff,Configuration.Geometry.GeometryExtended2017devReco_cff" --eventcontent FEVTDEBUGHLT --filein file:step1.root  --fileout file:step2.root --customise SLHCUpgradeSimulations/Configuration/HCalCustoms.customise_Hcal2017 > step2_SinglePiE50_UP15+SinglePiE50_UP15+DIGIUP15+RECOUP15+HARVESTUP15.log  2>&1

cmsDriver.py step3  --runUnscheduled  --conditions auto:phase1_2017_design -s RAW2DIGI,L1Reco,RECO,EI,PAT --datatier GEN-SIM-RECO,MINIAODSIM -n 10 --era Run2_2017 --geometry "Configuration.Geometry.GeometryExtended2017dev_cff,Configuration.Geometry.GeometryExtended2017devReco_cff" --eventcontent RECOSIM,MINIAODSIM --filein file:step2.root  --fileout file:step3.root --customise SLHCUpgradeSimulations/Configuration/HCalCustoms.customise_Hcal2017 > step3_SinglePiE50_UP15+SinglePiE50_UP15+DIGIUP15+RECOUP15+HARVESTUP15.log  2>&1
```

The problem(s) with `validRaw` cause a segmentation violation in `CaloTowersCreationAlgo::hadShwPosFromCells(DetId, DetId, float)` in step3.

To see which cells cause the problem, I added a cout near the end of the preceding function, `CaloTowersCreationAlgo::hadShwrPos(CaloTowerDetId towerId, float fracDepth)`:
```
+  std::cout << "CaloTowersCreationAlgo: front = " << frontCellId << ", back = " << backCellId << std::endl;
+
   point = hadShwPosFromCells(DetId(frontCellId), DetId(backCellId), fracDepth);
 
   return point;
```

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@smuzaffar @davidlange6 The compilation failure is not due to this PR. I believe there are issues in I for some of the the analysis codes

**Comment by kpedro88** [comment]
@bsunanda I think your latest change is just papering over the issue. The underlying problem is that depthBinInformation is returning the wrong value. We should make sure this works correctly.

https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/Geometry/CaloTopology/src/CaloTowerConstituentsMap.cc#L108

Here is a printout of the current behavior:
CaloTowerConstituentsMap: ieta = 17, sd = 2, nd = 3

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@kpedro88 Thanks for pointing out the problem. Hope it works ok now

**Comment by kpedro88** [comment]
@bsunanda It appears to be working correctly now, thanks!

**Comment by bsunanda** [comment]
@ianna @civanch @slava77 Could you kindly approve this

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/14594	427eede
- code changes in RECO are minimal: added a check for a valid ID in CaloTowersCreationAlgo.cc
- jenkins tests pass and comparisons with baseline show no differences

File: Geometry/CaloTopology/src/HcalTopology.cc [hunk 1]
```diff
@@ -144,7 +145,7 @@ HcalTopology::HcalTopology(HcalTopologyMode::Mode mode, int maxDepthHB, int maxD
   firstHETripleDepthRing_((mode==HcalTopologyMode::H2 || mode==HcalTopologyMode::H2HE)?(24):(27)),
   singlePhiBins_(72), doublePhiBins_(36),
```
Comment (ianna) [inline Geometry/CaloTopology/src/HcalTopology.cc]: @bsunanda - why singlePhiBins_ initialisation is kept unchanged here, while it was changed in the other constructor? BTW, the variable is not modified anywhere. Is it needed as a variable or the IPHI_MAX constant can be used instead?

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by ianna** [comment]
+1

**Comment by kpedro88** [comment]
@slava77 can you sign again? This one is very important for HCAL 2017 development.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/14594	1fbcfe7
- changes since last reco approval do not affect reco code
- jenkins tests still show no differences (as expected)

**Pull request merged**